src/qibo/derivative.py

```diff
@@ -0,0 +1,117 @@
+import numpy as np
+
+from qibo.config import raise_error
+from qibo.hamiltonians.abstract import AbstractHamiltonian
+
+
+def parameter_shift(
+    circuit,
+    hamiltonian,
+    parameter_index,
+    initial_state=None,
+):
+    """In this method the parameter shift rule (PSR) is implemented.
+    Given a circuit U and an observable H, the PSR allows to calculate the derivative
+    of the expected value of H on the final state with respect to a variational
+    parameter of the circuit.
+
+    Original references:
+        `https://arxiv.org/abs/1811.11184`;
+        `https://arxiv.org/abs/1803.00745`.
+
+    Args:
+        circuit (:class:`qibo.models.circuit.Circuit`): custom quantum circuit
+        hamiltonian (:class: `qibo.hamiltonians.Hamiltonian`): target observable
+        parameter_index (int): the index which identifies the target parameter in the circuit.get_parameters() list
+        initial_state ((1, 2**nqubits) matrix): initial state on which we act with the circuit.
+
+    Returns:
+        np.float value of the derivative of the expected value of H on the final state obtained applying U
+        to the initial state with respect to the target variational parameter.
+
+    Example:
+        .. testcode::
+
+            import qibo
+            import numpy as np
+            from qibo import hamiltonians, gates
+            from qibo.models import Circuit
+            from qibo.derivative import parameter_shift
+
+            # defining an observable
+            def hamiltonian(nqubits = 1):
+                m0 = (1/nqubits)*hamiltonians.Z(nqubits).matrix
+                ham = hamiltonians.Hamiltonian(nqubits, m0)
+                return ham
+
+            # defining a dummy circuit
+            def circuit(nqubits = 1):
+                c = Circuit(nqubits = 1)
+                c.add(gates.RY(q = 0, theta = 0))
+                c.add(gates.RX(q = 0, theta = 0))
+                c.add(gates.M(0))
+                return c
+
+            # initializing the circuit
+            c = circuit(nqubits = 1)
+
+            # some parameters
+            test_params = np.random.randn(2)
+            c.set_parameters(test_params)
+
+            test_hamiltonian = hamiltonian()
+
+            # running the psr with respect to the two parameters
+            grad_0 = parameter_shift(circuit = c, hamiltonian = test_hamiltonian, parameter_index = 0)
+            grad_1 = parameter_shift(circuit = c, hamiltonian = test_hamiltonian, parameter_index = 1)
+
+            print('Test gradient with respect params[0]: ', grad_0.numpy())
+            print('Test gradient with respect params[0]: ', grad_1.numpy())
+    """
```

Comment:
Question: I haven't tested this example, but will `.numpy()` work with all backends?

Reply:
Also I think it should be "test_params". In particular, "test_params[1]" in the second line.

Reply:
You are right, thank you. I put it in the first version of the example (when I used the tensorflow backend) but now it is not needed.